File: Array/linchuanxin/Algorithm/day32_0702_109有序链表转换二叉搜索树/day32_leetcode109.java
package day32;

public class day32_leetcode109 {



     public class ListNode {
          int val;
          ListNode next;
          ListNode(int x) { val = x; }
      }
      public class TreeNode {
          int val;
          TreeNode left;
          TreeNode right;
          TreeNode(int x) { val = x; }
     }

    class Solution {

        private ListNode findMiddleElement(ListNode head) {
            ListNode pre = null;
            ListNode fast = head;
            ListNode slow = head;

            while (fast != null && fast.next!=null){
                pre = slow;
                slow = slow.next;
                fast = fast.next.next;
            }

            if(pre!=null){
                pre.next = null;
            }

            return slow;

        }
        public TreeNode sortedListToBST(ListNode head) {
            if(head==null)
                return null;
            ListNode midNode = findMiddleElement(head);
            TreeNode node = new TreeNode(midNode.val);
            if(head == midNode){
                return node;
            }
            node.left = sortedListToBST(head);
            node.right = sortedListToBST(midNode.next);
            return node;
        }
    }
}

File: Array/linchuanxin/Algorithm/day32_0702_109有序链表转换二叉搜索树/day32_leetcode109.kt
package day32


  class ListNode(var `val`: Int) {
      var next: ListNode? = null
  }


  class TreeNode(var `val`: Int) {
      var left: TreeNode? = null
      var right: TreeNode? = null
  }

class Solution {

    fun findMidNode(head: ListNode?):ListNode?{
        var pre:ListNode? = null
        var fast:ListNode? = head
        var slow:ListNode? = head
        while (fast !=null && fast.next!=null){
            pre = slow
            slow = slow?.next;
            fast = fast.next?.next
        }
        if(pre != null){
            pre.next = null
        }
        return slow
    }


    fun sortedListToBST(head: ListNode?): TreeNode? {
        if (head == null)
            return null
        var midNode:ListNode? = findMidNode(head)
        var treeNode:TreeNode = TreeNode(midNode!!.`val`)
        if(head == midNode){
            return treeNode
        }
        treeNode.left = sortedListToBST(head)
        treeNode.right = sortedListToBST(midNode.next)
        return treeNode
    }
}